### PR TITLE
User functionality

### LINKED ISSUE #25: Adding functionality  to user model
- [x] Adding functionality to user model[create/delete/get].
- [x] Adding new file called test_user.py for testing user functions.
- [x] Added tests/generate_objects.py to generate all the objects we need.
- [ ] Adding user articles by likes/views.

### PR DESCRIPTION
**Pull request by rawadabu**
resolve #25 
-  Adding functionality to user model[create/delete/get].
-  Adding new file called test_user.py for testing user functions.
-  Added tests/generate_objects.py to generate all the objects we need.
-  Adding user articles by likes/views.

### CODE REVIEW

File: projboard/models.py
Comment (Shalevro2) [inline projboard/models.py]: 1. Should be:

`user = User(email=email,...)`
And not:
`User.objects.create_user(email=email, password=password, name=name, nickname=nickname)`

2. Consider to check the unique of nickname and email.

**Review by Shalevro2 — CHANGES_REQUESTED** [review]

**Review by Shalevro2 — APPROVED** [review]
LGTM

File: projboard/models.py
Comment (Shalevro2) [reply projboard/models.py]: I seen the last commit, ok for me.

File: projboard/models.py
Comment (avi-biton) [inline projboard/models.py]: Why do we check for both nickname and email?  
isn't one check enough? (get user either by mail or by nickname)

File: projboard/models.py
Comment (avi-biton) [inline projboard/models.py]: Since the method is not only creating a new User, but also checks if the user already exists and return it, I would rename the method's name to `get_user`

File: projboard/tests/test_user.py
Comment (avi-biton) [inline projboard/tests/test_user.py]: Why do we need the second user? it is not been used in the tests.

File: projboard/tests/test_user.py
Comment (Yarboa) [inline projboard/tests/test_user.py]: I think the tests should search it in database, isnt it?

**Review by Yarboa — CHANGES_REQUESTED** [review]

File: projboard/models.py
Comment (Yarboa) [reply projboard/models.py]: +1 here

File: projboard/models.py
Comment (Yarboa) [reply projboard/models.py]: +1

File: projboard/tests/test_user.py
Comment (Yarboa) [reply projboard/tests/test_user.py]: +1
In all test below, you are using only one generate_user[0]

I think you should also search for the user in db with the key field, or model method, isnt it? 

the search can be taken from fixture data

File: projboard/tests/test_user.py
Comment (finziamit) [inline projboard/tests/test_user.py]: wouldn't user1 be more readable as `User.create_user(name=NAME, nickname=NICKNAME, email=EMAIL, password=PASSWORD)` ?

**Review by Yarboa — CHANGES_REQUESTED** [review]

File: projboard/tests/test_user.py
Comment (Shalevro2) [reply projboard/tests/test_user.py]: @Yarboa You mean to something like that:
```
@pytest.mark.django_db
def test_create_user(self, generate_user):
        assert generate_user in User.objects.all()
```

?

File: projboard/models.py
Comment (rawadabu) [reply projboard/models.py]: Done

File: projboard/models.py
Comment (rawadabu) [reply projboard/models.py]: Done

File: projboard/tests/test_user.py
Comment (rawadabu) [reply projboard/tests/test_user.py]: Done

File: projboard/tests/test_user.py
Comment (rawadabu) [reply projboard/tests/test_user.py]: Done

File: projboard/tests/test_user.py
Comment (Yarboa) [reply projboard/tests/test_user.py]: This could be one option, the other query a method with user fixture name,
retrieve from db, assert they are equal

Comment (rawadabu) [reply projboard/tests/test_user.py]: @Yarboa Can you review now please?

Comment (Yarboa) [reply projboard/tests/test_user.py]: Almost 
   @pytest.mark.django_db()
    def test_create_user(self, generate_user):
        generate_user.save()
        assert generate_user in User.objects.all()
        
 Please move the save to fixture

**Review by Yarboa — COMMENTED** [review]
Nice work

Please refer this
https://github.com/redhat-beyond/Smarticle/pull/41#discussion_r1036023029

**Comment by rawadabu** [comment]
> Nice work
> 
> Please refer this [#41 (comment)](https://github.com/redhat-beyond/Smarticle/pull/41#discussion_r1036023029)

@Yarboa Done.

**Review by Shalevro2 — APPROVED** [review]
Great job, LGTM.

**Review by yaelnoyman1 — APPROVED** [review]
LGTM

**Review by Anas-mhameed — APPROVED** [review]
LGTM

**Review by finziamit — APPROVED** [review]
LGTM

File: projboard/tests/test_user.py
Comment (Yarboa) [inline projboard/tests/test_user.py]: I assume you should use this 
User.get_user(generate_user.email,  generate_user.nickname, generate_user.name, generate_user.nickname):

And then assert it?

Comment (rawadabu) [reply projboard/tests/test_user.py]: @Yarboa Done

Comment (Yarboa) [reply projboard/tests/test_user.py]: But you have to assert the User content with user_generated data,
You should compare what the test prepared with what was generated in DB
Dont you think?

**Review by Yarboa — COMMENTED** [review]
Please rebase main 
resolve conflict and push --force

**Comment by rawadabu** [comment]
> Please rebase main
> resolve conflict and push --force

@Yarboa Done

File: projboard/tests/test_user.py
Comment (rawadabu) [reply projboard/tests/test_user.py]: @Yarboa Done

File: projboard/tests/test_user.py
Comment (avi-biton) [inline projboard/tests/test_user.py]: The test name indicates that you are testing creation of a new user.  
However, the test uses `get_user_by_nickname` and not creating a new user.

You should name your tests according to the method they test.
For example:
If you want to test `get_user`, the test name should be `test_get_user` and you can create multiple tests for the same method using informative names like:
- `get_user_new_user`
- `get_user_existing_user`
- `get_user_fail`

etc...

File: projboard/tests/test_user.py
Comment (avi-biton) [inline projboard/tests/test_user.py:40]: No need to write a different test for it, you can add it to the previous test

File: projboard/tests/test_user.py
Comment (avi-biton) [inline projboard/tests/test_user.py]: The test name should be renamed to `test_get_user_by_nickname`

File: projboard/tests/test_user.py
Comment (avi-biton) [inline projboard/tests/test_user.py]: The test only tests that the user is in the DB, but it does not check if the user that returned from User.get_user_by_nickname(generate_user.nickname) is the user we want.
You did these assertions exactly in the first test

File: projboard/models/user.py
Comment (avi-biton) [inline projboard/models/user.py:48]: You would never get this exception, because it is not raised anywhere.  
`get_user_by_nickname` is not rising anything, it catches the `ObjectDoesNotExist` and returns `None`

You should either change `get_user_by_nickname` to raise the exception or change your implementation

**Comment by avi-biton** [comment]
@rawadabu the method `get_user_by_nickname` is defined in the file, but its not been used anywhere.  
I see you are using `User.objects.get(nickname=nickname)` instead.  

So either remove the `get_user_by_nickname`, or change the calls you use.

**Comment by rawadabu** [comment]
> @rawadabu the method `get_user_by_nickname` is defined in the file, but its not been used anywhere. I see you are using `User.objects.get(nickname=nickname)` instead.
> 
> So either remove the `get_user_by_nickname`, or change the calls you use.

Got it, I updated User functions to use `User.get_user_by_nickname(nickname=nickname)` instead, except in `get_user_by_nickname` function I used `User.objects.get(nickname=nickname)`

File: projboard/tests/test_user.py
Comment (Yarboa) [inline projboard/tests/test_user.py]: Please remove the comments

File: projboard/tests/test_user.py
Comment (Yarboa) [reply projboard/tests/test_user.py]: +1

File: projboard/tests/test_user.py
Comment (Yarboa) [reply projboard/tests/test_user.py]: +1

File: projboard/tests/test_user.py
Comment (Yarboa) [reply projboard/tests/test_user.py]: +1

**Review by Yarboa — COMMENTED** [review]
Please rebase main 
resolve conflict and push --force

File: projboard/tests/test_user.py
Comment (rawadabu) [reply projboard/tests/test_user.py]: Done

File: projboard/tests/test_user.py
Comment (rawadabu) [reply projboard/tests/test_user.py]: Done

File: projboard/tests/test_user.py
Comment (rawadabu) [reply projboard/tests/test_user.py]: I assume that I've done it right

Comment (rawadabu) [reply projboard/tests/test_user.py]: Updated and rebased with main once again (after merging Shalev's PR)
Waiting for your review @avi-biton @Yarboa

File: projboard/models/user.py
Comment (avi-biton) [inline projboard/models/user.py]: Do not leave unfinished code, either remove it or remark it

Comment (rawadabu) [reply projboard/models/user.py]: @avi-biton Done

File: projboard/tests/test_user.py
Comment (rawadabu) [reply projboard/tests/test_user.py]: Done

File: projboard/tests/test_user.py
Comment (rawadabu) [reply projboard/tests/test_user.py:40]: Done

File: projboard/models/user.py
Comment (Shalevro2) [inline projboard/models/user.py]: Consider to comment this lines also.

File: projboard/models/user.py
Comment (Shalevro2) [inline projboard/models/user.py]: Also here

**Review by Shalevro2 — CHANGES_REQUESTED** [review]
The tests are looking good to me.

**Review by Shalevro2 — APPROVED** [review]
LGTM

**Pull request merged**
